[DROOLS-729] Android support

kedzie opened the pull request:
-drools-android module added
-dex classloaders for generated classes
-maven plugin for pre-serializing knowledge packages during buildtime
-roboguice extension for injecting kiebases from pre serialized packages
-Kie classpath container works.  loads compilation cache from maven build
-added drools-android-examples. one with preserialized kbases and one with classpath container

At this point the only advantage of using the preserialized kbases is that it requires lot less dependencies.  With the drools-compiler dependency I needed to use multidex to package all the classes.  Which some users might not want.. to limit the size of the apk.

-This change also depends on a change to the maven-android-plugin, to which I have already submitted a pull request.  
-Finally android 5.0 support depends on a change to mvel2, namely it's handling of the java.version system property, which cannot be overridden in android 5.0.  I am also submitting a pull request for that.

Looking forward to feedback!

mrietveld commented:
Maybe this should be added to the [droolsjbpm-integration](https://github.com/droolsjbpm/droolsjbpm-integration) repository instead of the the drools repository?

kedzie commented:
Well this functionality includes code changes within drools-core, drools-compiler, and kie-maven-plugin to support the functionality.  Mainly it detects the Android environment and loads alternate classloader implementations (which live in drools-android).   

Though i agree it might make sense to move the drools-android module itself over to the integration repository.

kedzie commented:
The mvel PR has been submitted to make this work under Android 5.0

mrietveld commented:
@kedzie One more question: I couldn't find the license for the android:android artifact. I also couldn't find a repository that it was available in. 

Could you point me to the project page, especially the license? Also, if you can point me to an online repository (such as https://repo.maven.apache.org/ or https://repository.jboss.org/) where the artifact is available, that would be great.

mrietveld commented on drools-android/drools-android.iml:
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
nit-picking: this file should also be removed.

mrietveld commented on drools-android/src/main/java/org/drools/android/DroolsAndroidContext.java:
```diff
@@ -0,0 +1,274 @@
+    public static void setContext(Context ctx) {
+        context = ctx;
+        cacheDir = new File(context.getCacheDir(), "drools");
+//        if (cacheDir != null && cacheDir.isDirectory()) {
```
If you could delete commented out code, that would be great.

mrietveld commented on drools-examples-android/kiecontainer/pom.xml:
```diff
@@ -0,0 +1,233 @@
+            <groupId>org.kie</groupId>
+            <artifactId>kie-internal</artifactId>
+        </dependency>
+        <!--<dependency>-->
```
Please remove commented out code. Thanks!

mrietveld commented on maven_repo_3rd_party/android/android/4.4.2_r3/_remote.repositories:
```diff
@@ -0,0 +1,6 @@
+#NOTE: This is an Aether internal implementation file, its format can be changed without prior notice.
```
This file should probably be removed?

mrietveld commented:
@kedzie It also looks like the "maven_repo_3rd_party/" files were added by accident. 

Could you confirm that they should not be part of the pull-request -- or otherwise, could you explain why they need to be added to the pull-request? 

Thanks!

mrietveld commented:
FYI: not sure how familiar you are with git/github. In order to modify a pull-request, you can either push new commits or push -f to your pull-request branch, and they will be added to the pull-request.

kedzie commented:
A copy of the license can be found here: https://dl-ssl.google.com/android/repository/repository-10.xml .

The android platform jars are not available in any published maven repo.  They come as part of the android SDK.  There are subfolders in the sdk which are actually maven repositories.  There is also an open source project called maven-android-sdk-deployer which deploys artifacts from a local Android SDK installation to your local repo.  This is why i including the maven_3rd_party folder.  So there wouldn't be any unresolved dependencies.  

Not sure why Google doesn't publish the latest versions of android to public repos (they did publish the first few versions to maven centrel).  If you know a better solution to this let me know.

kedzie commented:
Will remove the commented code.

mrietveld commented:
Sorry for the late reply (and thank you for your patience!). I would really like to get this pull request in because I think lots of people would appreciate it. 

Is there a way that we can add the code without adding the "maven_repo_3rd_party/" directory? I can't see us adding those files to a core repository. Any and all idea's are welcome (ant/maven task to download the files and create the directory?). I know how to make a maven task to download files -- but is there a.. default/standard URL for downloading the android SDK?

kedzie commented:
There is, and it can definately be automated.  The Android SDK is also required to build the drools android example projects included in this PR.  Also the ${ANDROID_HOME} environment variable needs to be set properly.  The Android SDK also includes maven repo's as extra components (which can be downloaded into the SDK via command line).

kedzie commented:
Something like this can be done to reference the m2repo under the Android SDK.  

 <repositories>
<repository>
         <id>com.android.support</id>
         <url>file://${env.ANDROID_HOME}/extras/android/m2repository</url>
     </repository>
</repositories>

However there is also a jar I depend on that isn't in the android SDK maven repo.  It is in the SDK however, in the build-tools folder.  it is called dx.jar.  It is the .class -> .dex converter tool.

kedzie commented:
There are many examples of projects that depend on android artifacts.  Such as [maven-android-plugin](https://github.com/simpligility/android-maven-plugin).  They have [travis-CI configurations](https://github.com/simpligility/android-maven-plugin/blob/master/.travis.yml) which download the android-sdk and required components.  Even launch an emulator to perform integration testing.  Which is a great way to test this functionality by the way, to use the drools-android example projects I included as integration tests to verify the drools functionality is working.

kedzie commented:
The main android jar, version 4.1.1.4 is available in maven central as [found here](http://mvnrepository.com/artifact/com.google.android/android).  Which I believe is sufficient for our needs.   This just leaves the dx jar.

kedzie commented:
The DX library I need is available, in part, in maven central under apache license with the name com.google.dexmaker:dexmaker-dx:1.0.0.  It isn't the complete library however, but a subset.  But it is under apache license which is good news.  Then I for examples I use the multidex libraries, which are in the android SDK m2repos.

kedzie commented:
finally the apache-harmony javax.beans library.  I made the package myself by jarjar'ing the harmony jar to rip out the javax.beans classes only.  Android doesn't have the javax.beans introspection stuff.

mrietveld commented:
Hi @kedzie, 

We got a response back from legal with regards to the code: 

1. It looks like Android SDK-associated jars are ASL 2.0, which is good. 
2. There's a copyright header in a number of pieces of code that starts with  
`* * Copyright (C) 2014 CNH Industrial NV. All rights reserved.`
We're ASL 2.0, we can't include any code with this copyright. 

Could you give me written confirmation that removing this header from your code would be okay? 
(or otherwise remove the headers yourself?). 

Lastly, sorry for the delayed reply -- it's been kind of hectic for us. It's a pretty busy period at the moment.

kedzie commented:
Sorry about the copyright. That is my work copyright. Didn't mean to have
that there. I will remove it.
On Apr 8, 2015 5:20 AM, "Marco Rietveld" <notifications@github.com> wrote:

> Hi @kedzie <https://github.com/kedzie>,
>
> We got a response back from legal with regards to the code:
>
>    1. It looks like Android SDK-associated jars are ASL 2.0, which is
>    good.
>    2. There's a copyright header in a number of pieces of code that
>    starts with
>    * * Copyright (C) 2014 CNH Industrial NV. All rights reserved. We're
>    ASL 2.0, we can't include any code with this copyright.
>
> Could you give me written confirmation that removing this header from your
> code would be okay?
> (or otherwise remove the headers yourself?).
>
> Lastly, sorry for the delayed reply -- it's been kind of hectic for us.
> It's a pretty busy period at the moment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/droolsjbpm/drools/pull/416#issuecomment-90872365>.
>

psiroky commented:
Hello @kedzie. I will look into the changes and will work with you to get this merged.

As a first thing, could you please split the PR into two and move the `drools-android` and `drools-examples-android` into droolsjbpm-integration repo, which I believe is the most correct place for them.

Only the changes to `drools-core` (and other Drools modules should stay here). Let me know if you have any questions. Thanks!

kedzie commented:
Yes I will move and create a new pull request in drools-integration

kedzie commented:
Also are you aware of what all this change depends on?   And what needs to happen to get this merged?  Here is a list:

* mvel/mvel#72] needs to get merged
* droolsjbpm/droolsjbpm-knowledge#84 needs to get merged
* Dependencies need to be resolved. Here is the list of 3rd party dependencies not found in maven central (currently found in maven_repo_3rd_party folder).  They can be published to a JBoss Maven Repository or I have also written a script called [install-dependencies.sh](https://github.com/kedzie/drools-android/blob/master/drools-android/install-dependencies.sh) which installs the android dependencies from the Android SDK into the local maven repo.
  * Apache Harmony javax.beans package.  I stripped the javax.beans package from the harmony jars and packaged them into a jar file.  Android OS doesn't have that package.  If there is another existing jar in maven central which has these that could be used.  Otherwise this needs to be deployed to some JBoss Maven Repository.
  * Android DX library (DEX Converter) version 1.10.  I packaged this from Android Open Source repositories.  It is also available in the Android SDK at *$ANDROID_HOME/build-tools/21.1.2/lib/dx.jar*.
  * Android multidex and multidex-instrumentation libraries.  They are both found in the Android SDK extra maven repositories.

psiroky commented:
Yes, I am aware of the other PRs, will look at them too.

As for the dependencies, in ideal case everything should be accessible from some online repository. If that is not possible for whatever reason, we need to figure out the best possible way to include them (I am not sure what would that be right now).

I would like to take smaller steps here as the PR itself is quite big and it would be very hard to resolve everything at once. So once you split the PR properly we can focus on the other things. Does that sound good?

kedzie commented:
Moved drools-android & drools-examples-android to droolsjbpm/droolsjbpm-integration#118.  So in this project it is the environment detection, Classloader instantiation, and the kie-maven-plugin modifications to preserialize kiebases.

psiroky commented:
Can you please elaborate a bit more about the `kie-maven-plugin` changes? Are they just Android specific (e.g. not needed outside Android env)? Are there any implications for non-Android users? We need to make sure any Android-related change will not impact the usual use cases.

kedzie commented:
I added a goal for preserializing the kiebases.  It compiles the specified kiebases and then serializes them to filesystem.  I use it for Android so it can execute without drools-compiler dependency.  It doesn't reference Android in any way.  I think I also added a test case as well.

kedzie commented:
It is a separate goal.  There were no changes made to the existing goals in the plugin.  So it shouldn't have any effect on anything.  The test case is a kie-maven-plugin example that existed before, I just moved it into an maven integration test.

psiroky commented:
@kedzie thanks for the info. Since it is just an optional goal, it really should not affect the "standard" usage. I will take a closer look tomorrow or on Friday.

viridian1138 commented:
I'm curious about the lack of a res/layout folder in the sample code.  It seems as if the entire operation of the sample code is to be verified/evaluated through the android logging service.  The addition of a layout with a GUI would make it easier to demo the code, and would make it easier use the sample code with IDEs like AIDE that really don't work well with the idea of running the app while simultaneously watching the logging console.

kedzie commented:
@viridian1138  Which sample code are you referring too?  The kie-maven-plugin integration test?  That is not an android project.  It is just a java project.  The android examples are in drools-integration.  They do have gui's and automated testing.

kedzie commented:
Rebased.  Also this depends on an MVEL pull request mve/mvel#72 to get released.  It believe it got merged, but it needs to be released so we can reference the new mvel version in the pom.

psiroky commented:
@kedzie mvel 2.2.5.Final has been released with your change in. It is in JBoss repo (repository.jboss.org). For some reason it has not been synced into Maven central, we will need to fix that.

mariofusco commented on drools-core/src/main/java/org/drools/core/util/ClassUtils.java:
```diff
@@ -621,6 +672,20 @@ public static boolean isOSX() {
+     * Checks if running on Android operating system
+     * @return
+     */
+    public static boolean isAndroid() {
```
This method is invoked many times and I believe that after it has been called the first time its result could be cached forever since those classes either are present in the classpath or aren't.

mariofusco commented on drools-core/src/main/java/org/drools/core/util/Drools.java:
```diff
@@ -18,7 +18,7 @@
     static {
         droolsFullVersion = Drools.class.getPackage().getImplementationVersion();
-        if (droolsFullVersion == null) {
+        if (droolsFullVersion == null || droolsFullVersion.equals("0.0")) {
```
In which case does getImplementationVersion() return "0.0". Is that an android behaviour? Is it documented?

mariofusco commented on drools-core/src/main/java/org/drools/core/util/IoUtils.java:
```diff
@@ -125,6 +125,7 @@ public static void copyFile(File sourceFile, File destFile) {
             Enumeration< ? extends ZipEntry> entries = zipFile.entries();
             while ( entries.hasMoreElements() ) {
                 ZipEntry entry = entries.nextElement();
+                if(entry.getName().endsWith(".dex")) continue; //avoid out of memory error, it is useless anyway
```
Please follow Drools code convention and put the continue; in a new line and between { }

kedzie replied:
Will do
On Jun 5, 2015 3:31 AM, "Mario Fusco" <notifications@github.com> wrote:

> In drools-core/src/main/java/org/drools/core/util/IoUtils.java
> <https://github.com/droolsjbpm/drools/pull/416#discussion_r31795504>:
>
> > @@ -125,6 +125,7 @@ public static void copyFile(File sourceFile, File destFile) {
> >              Enumeration< ? extends ZipEntry> entries = zipFile.entries();
> >              while ( entries.hasMoreElements() ) {
> >                  ZipEntry entry = entries.nextElement();
> > +                if(entry.getName().endsWith(".dex")) continue; //avoid out of memory error, it is useless anyway
>
> Please follow Drools code convention and put the continue; in a new line
> and between { }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/droolsjbpm/drools/pull/416/files#r31795504>.
>

kedzie commented:
I will cache the isAndroid().  And yes the implementation version returns 0. I can try to find android java docs for it. @mariofusco

kedzie commented:
@mariofusco isAndroid() is already cached.  Those class lookups are cached in a HashMap.  You want to cache it further?

kedzie commented:
@mariofusco  Since mvel 2.2.5 got released, can we update drools to use the new version?   The version is set in a parent pom, so it cannot be done in this change.

psiroky commented:
@kedzie I just upgraded the mvel to 2.2.5.Final in our parent. https://github.com/droolsjbpm/droolsjbpm-build-bootstrap/commit/d071a89417d587972c7835ac7e349e81394a4a72

kedzie commented:
@psiroky I merged in your PR and squashed them in.

psiroky commented:
@kedzie thanks. I know about the changes. I will merge this PR through out the weekend. My apologies for taking so long to merge this.

psiroky commented:
Merged into master https://github.com/droolsjbpm/drools/commit/c8e09e2056c54ead97bce4386a25b222154223b1. Now one last PR remaining :)

kedzie commented:
Great news!
